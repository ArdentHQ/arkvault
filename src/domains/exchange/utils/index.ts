export * from "./assertions";
export * from "./errors";
