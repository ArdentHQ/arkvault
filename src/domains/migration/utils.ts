import { MigrationTransactionStatus } from "@/domains/migration/migration.contracts";

export const getIcon = (status: MigrationTransactionStatus) => {
	if (status === MigrationTransactionStatus.Confirmed) {
		return {
			color: "text-theme-success-600",
			name: "CircleCheckMark",
		};
	}

	return {
		color: "text-theme-warning-300",
		name: "Clock",
	};
};
