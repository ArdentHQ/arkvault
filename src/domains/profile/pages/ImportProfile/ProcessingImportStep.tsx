import { Contracts, Environment } from "@ardenthq/sdk-profiles";
import React, { useEffect, useState } from "react";
import { useTranslation } from "react-i18next";

import { ReadableFile } from "@/app/hooks/use-files";
import { FilePreview } from "@/domains/profile/components/FilePreview";
import { PasswordModal } from "@/domains/profile/components/PasswordModal";
import { useProfileImport } from "@/domains/profile/hooks/use-profile-import";
import { Alert } from "@/app/components/Alert";
import { Button } from "@/app/components/Button";
import { FormButtons } from "@/app/components/Form";
import { StepHeader } from "@/app/components/StepHeader";
import { useNavigationContext } from "@/app/contexts";
import { Icon } from "@/app/components/Icon";

interface ImportErrorProperties {
	file: ReadableFile;
	onCancel: () => void;
	onRetry: () => void;
}

const ImportError = ({ file, onCancel, onRetry }: ImportErrorProperties) => {
	const { t } = useTranslation();
	const { setShowMobileNavigation } = useNavigationContext();

	useEffect(() => {
		// Stick form buttons to bottom.
		setShowMobileNavigation(false);
	}, []);

	return (
		<div className="mx-auto max-w-xl" data-testid="ImportError">
			<Alert className="mt-4" variant="danger">
				{t("PROFILE.IMPORT.PROCESSING_IMPORT_STEP.ERROR")}
			</Alert>

			<div className="mt-4">
				<FilePreview file={file} variant="danger" />
			</div>

			<FormButtons>
				<Button data-testid="ImportError__cancel" variant="secondary" onClick={onCancel}>
					{t("COMMON.CANCEL")}
				</Button>

				<Button data-testid="ImportError__retry" onClick={onRetry}>
					{t("COMMON.RETRY")}
				</Button>
			</FormButtons>
		</div>
	);
};

interface ProcessingImportProperties {
	env: Environment;
	file: ReadableFile;
	password?: string;
	shouldRequestPassword?: boolean;
	onBack: () => void;
	onCancel: () => void;
	onPasswordChange: (password?: string) => void;
	onRetry: () => void;
	onSuccess: (profile: any) => void;
}

export const ProcessingImport = ({
	env,
	file,
	password,
	shouldRequestPassword = false,
	onBack,
	onCancel,
	onPasswordChange,
	onRetry,
	onSuccess,
}: ProcessingImportProperties) => {
	const { t } = useTranslation();
	const { importProfile } = useProfileImport({ env });

	const [hasError, setHasError] = useState(false);
	const [isPasswordModalOpen, setIsPasswordModalOpen] = useState(shouldRequestPassword);

	useEffect(() => {
		const runImport = async () => {
			let profile: Contracts.IProfile | undefined;

			try {
				profile = await importProfile({ file, password });
				onSuccess(profile);
			} catch (error) {
				if (error.message === "PasswordRequired") {
					setIsPasswordModalOpen(true);
					return;
				}

				setHasError(true);
			}
		};

		if (file) {
			runImport();
		}
	}, [file, env, password]); // eslint-disable-line react-hooks/exhaustive-deps

	const renderStep = () => {
		if (hasError) {
			return <ImportError file={file} onCancel={onCancel} onRetry={onRetry} />;
		}

		return <FilePreview file={file} variant="loading" />;
	};

	return (
		<div className="mx-auto max-w-xl" data-testid="ProcessingImport">
			<StepHeader
				titleIcon={<Icon name="ImportProfile" dimensions={[24, 24]} />}
				title={t("PROFILE.IMPORT.TITLE")}
				subtitle={t("PROFILE.IMPORT.PROCESSING_IMPORT_STEP.DESCRIPTION", { name: file?.name || "" })}
			/>

			<div className="mt-4">{renderStep()}</div>

			<div className="items-left text-left">
				<PasswordModal
					isOpen={isPasswordModalOpen}
					title={t("PROFILE.IMPORT.PASSWORD_TITLE")}
					description={t("PROFILE.IMPORT.PASSWORD_DESCRIPTION")}
					onSubmit={(password: string) => {
						setIsPasswordModalOpen(false);
						onPasswordChange(password);
					}}
					onClose={() => {
						setIsPasswordModalOpen(false);
						onPasswordChange(undefined);
						onBack();
					}}
				/>
			</div>
		</div>
	);
};
