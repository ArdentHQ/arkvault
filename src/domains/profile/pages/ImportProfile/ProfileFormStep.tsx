import React from "react";
import { useTranslation } from "react-i18next";

import { Contracts, Environment } from "@payvo/sdk-profiles";
import { Divider } from "@/app/components/Divider";
import { FilePreview } from "@/domains/profile/components/FilePreview";
import { ProfileForm, ProfileFormState } from "@/domains/profile/components/ProfileForm";
import { ReadableFile } from "@/app/hooks/use-files";
import { StepHeader } from "@/app/components/StepHeader";

interface ImportProfileFormProperties {
	file?: ReadableFile;
	profile: Contracts.IProfile;
	password?: string;
	env: Environment;
	shouldValidate: boolean;
	showPasswordFields?: boolean;
	onSubmit: (profile: Contracts.IProfile) => void;
	onBack: () => void;
}

export const ImportProfileForm: React.VFC<ImportProfileFormProperties> = ({
	profile,
	env,
	onSubmit,
	onBack,
	file,
	password,
	shouldValidate,
}) => {
	const { t } = useTranslation();

	const handleSubmit = async ({
		avatarImage,
		name,
		password: enteredPassword,
		currency,
		viewingMode,
	}: ProfileFormState) => {
		env.profiles().push(profile);
		await env.profiles().restore(profile, password);

		profile.settings().set(Contracts.ProfileSetting.Name, name.trim());
		profile.settings().set(Contracts.ProfileSetting.Theme, viewingMode);
		profile.settings().set(Contracts.ProfileSetting.Avatar, avatarImage);
		profile.settings().set(Contracts.ProfileSetting.ExchangeCurrency, currency);

		if (enteredPassword || password) {
			// @ts-ignore
			profile.auth().setPassword(enteredPassword || password);
		}

		onSubmit(profile);
	};

	return (
		<div className="mx-auto max-w-xl">
			<StepHeader title={t("PROFILE.IMPORT.TITLE")} subtitle={t("PROFILE.IMPORT.FORM_STEP.DESCRIPTION")} />

			<div className="rounded-lg border-theme-secondary-300 bg-theme-background dark:border-theme-secondary-800 sm:mt-6 sm:border sm:p-10">
				<div className="mt-6 sm:mt-0">
					<FilePreview file={file} variant="success" useBorders={false} />
				</div>

				<Divider />

				<ProfileForm
					defaultValues={{
						avatarImage: profile.settings().get(Contracts.ProfileSetting.Avatar, ""),
						confirmPassword: password,
						currency: profile.settings().get(Contracts.ProfileSetting.ExchangeCurrency),
						name: profile.name(),
						password,
						viewingMode: profile.settings().get(Contracts.ProfileSetting.Theme),
					}}
					onBack={onBack}
					onSubmit={handleSubmit}
					shouldValidate={shouldValidate}
					showPasswordFields={!profile.usesPassword()}
				/>
			</div>
		</div>
	);
};
