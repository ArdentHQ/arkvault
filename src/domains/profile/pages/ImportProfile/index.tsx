export { ImportProfile as default } from "./ImportProfile";
