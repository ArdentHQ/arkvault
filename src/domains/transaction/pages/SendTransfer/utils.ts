export const getFeeType = (numberOfRecipients: number): "multiPayment" | "transfer" =>
	numberOfRecipients > 1 ? "multiPayment" : "transfer";
