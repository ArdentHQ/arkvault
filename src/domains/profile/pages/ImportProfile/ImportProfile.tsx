import { Contracts } from "@payvo/sdk-profiles";
import React, { useState } from "react";
import { useTranslation } from "react-i18next";
import { useHistory } from "react-router-dom";

import { Page, Section } from "@/app/components/Layout";
import { TabPanel, Tabs } from "@/app/components/Tabs";
import { StepsProvider, useEnvironmentContext } from "@/app/contexts";
import { ReadableFile } from "@/app/hooks/use-files";
import { ProcessingImport } from "@/domains/profile/pages/ImportProfile/ProcessingImportStep";
import { ImportProfileForm } from "@/domains/profile/pages/ImportProfile/ProfileFormStep";
import { SelectFileStep } from "@/domains/profile/pages/ImportProfile/SelectFileStep";

enum Step {
	SelectFileStep = 1,
	ProcessingStep,
	FormStep,
}

export const ImportProfile = () => {
	const { env, persist } = useEnvironmentContext();
	const { t } = useTranslation();
	const history = useHistory();

	const [activeTab, setActiveTab] = useState<Step>(Step.SelectFileStep);
	const [fileFormat, setFileFormat] = useState(".wwe");
	const [selectedFile, setSelectedFile] = useState<ReadableFile>();
	const [password, setPassword] = useState<string>();
	const [profile, setProfile] = useState<Contracts.IProfile>();

	const handleSelectedFile = (file: ReadableFile) => {
		setSelectedFile(file);
		setActiveTab(Step.ProcessingStep);
	};

	const handleProfileSave = () => {
		persist();
		history.push("/");
	};

	return (
		<Page
			navbarVariant="logo-only"
			pageTitle={t("PROFILE.PAGE_WELCOME.IMPORT_PROFILE_TITLE")}
			title={t("COMMON.PAYVO_WALLET")}
		>
			<Section className="-mt-8">
				<StepsProvider activeStep={activeTab} steps={3}>
					<Tabs activeId={activeTab} className="mt-6 sm:mt-8">
						<TabPanel tabId={Step.SelectFileStep}>
							<SelectFileStep
								fileFormat={fileFormat}
								onFileFormatChange={setFileFormat}
								onSelect={handleSelectedFile}
								onBack={() => history.push("/")}
							/>
						</TabPanel>

						<TabPanel tabId={Step.ProcessingStep}>
							{selectedFile && (
								<ProcessingImport
									env={env}
									password={password}
									file={selectedFile}
									onBack={() => setActiveTab(Step.SelectFileStep)}
									onCancel={() => history.push("/")}
									onPasswordChange={setPassword}
									onRetry={() => {
										setSelectedFile({ ...selectedFile });
										setPassword(undefined);
									}}
									onSuccess={(profile) => {
										setProfile(profile);
										setActiveTab(Step.FormStep);
									}}
								/>
							)}
						</TabPanel>

						<TabPanel tabId={Step.FormStep}>
							{profile && (
								<ImportProfileForm
									file={selectedFile}
									env={env}
									profile={profile}
									password={password}
									shouldValidate
									onSubmit={handleProfileSave}
									onBack={() => {
										setPassword(undefined);
										setActiveTab(Step.SelectFileStep);
									}}
								/>
							)}
						</TabPanel>
					</Tabs>
				</StepsProvider>
			</Section>
		</Page>
	);
};
