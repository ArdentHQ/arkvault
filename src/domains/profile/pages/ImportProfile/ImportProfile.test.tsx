/* eslint-disable @typescript-eslint/require-await */
import fs from "fs";
import userEvent from "@testing-library/user-event";
import { createHashHistory } from "history";
import React from "react";

import { EnvironmentProvider } from "@/app/contexts";
import { ImportProfile } from "@/domains/profile/pages/ImportProfile/ImportProfile";
import { env, fireEvent, render, screen, waitFor } from "@/utils/testing-library";

const passwordProtectedWwe = fs.readFileSync("src/tests/fixtures/profile/import/password-protected-profile.wwe");
const corruptedWwe = fs.readFileSync("src/tests/fixtures/profile/import/corrupted-profile.wwe");
const legacyJson = fs.readFileSync("src/tests/fixtures/profile/import/legacy-profile.json");
const darkThemeWwe = fs.readFileSync("src/tests/fixtures/profile/import/profile-dark-theme.wwe");
const lightThemeWwe = fs.readFileSync("src/tests/fixtures/profile/import/profile-light-theme.wwe");
const history = createHashHistory();

const importProfileURL = "/profiles/import";

const browseFiles = () => screen.getByTestId("SelectFile__browse-files");

const changeFileID = "SelectFileStep__change-file";
const submitID = "PasswordModal__submit-button";
const validPassword = "S3cUrePa$sword";
const wrongPassword = "wrong password";

const createBlob = (fileContents: string | Buffer, fileName?: string) =>
	new File([new Blob([fileContents])], fileName || "fileName.wwe");

describe("ImportProfile", () => {
	beforeEach(() => {
		history.push(importProfileURL);
	});

	it("should render first step", async () => {
		render(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
		);

		expect(screen.getByTestId(changeFileID)).toBeInTheDocument();
	});

	it("should go back", async () => {
		const historyMock = vi.spyOn(history, "push").mockReturnValue();

		render(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(screen.getByTestId(changeFileID)).toBeInTheDocument();
		expect(screen.getByTestId("SelectFileStep__back")).toBeInTheDocument();

		await userEvent.click(screen.getByTestId("SelectFileStep__back"));

		await waitFor(() => expect(historyMock).toHaveBeenCalledWith("/"));
		historyMock.mockRestore();
	});

	it("should change file format", async () => {
		render(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(screen.getByTestId(changeFileID)).toBeInTheDocument();

		await userEvent.click(screen.getByTestId(changeFileID));

		await waitFor(() => expect(screen.queryByTestId(changeFileID)).not.toBeInTheDocument());
	});

	it("should select file and go to step 2", async () => {
		render(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(screen.getByTestId(changeFileID)).toBeInTheDocument();
		expect(screen.getByTestId("SelectFileStep__back")).toBeInTheDocument();

		fireEvent.drop(browseFiles(), {
			dataTransfer: {
				files: [createBlob(passwordProtectedWwe)],
			},
		});

		await expect(screen.findByTestId("ProcessingImport")).resolves.toBeVisible();
	});

	it("should request and set password for importing password protected profile", async () => {
		render(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(screen.getByTestId(changeFileID)).toBeInTheDocument();
		expect(screen.getByTestId("SelectFileStep__back")).toBeInTheDocument();

		fireEvent.drop(browseFiles(), {
			dataTransfer: {
				files: [createBlob(passwordProtectedWwe)],
			},
		});

		await expect(screen.findByTestId("ProcessingImport")).resolves.toBeVisible();
		await expect(screen.findByTestId("Modal__inner")).resolves.toBeVisible();

		await userEvent.type(screen.getByTestId("PasswordModal__input"), validPassword);

		// wait for formState.isValid to be updated
		await expect(screen.findByTestId(submitID)).resolves.toBeVisible();

		await userEvent.click(screen.getByTestId(submitID));

		await waitFor(() => {
			expect(screen.getByTestId("ProfileForm__form")).toBeInTheDocument();
		});
	});

	it("should close password modal and go back to select file", async () => {
		render(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(screen.getByTestId(changeFileID)).toBeInTheDocument();
		expect(screen.getByTestId("SelectFileStep__back")).toBeInTheDocument();

		fireEvent.drop(browseFiles(), {
			dataTransfer: {
				files: [createBlob(passwordProtectedWwe)],
			},
		});

		await expect(screen.findByTestId("ProcessingImport")).resolves.toBeVisible();
		await expect(screen.findByTestId("Modal__inner")).resolves.toBeVisible();

		await userEvent.click(screen.getByTestId("Modal__close-button"));

		await expect(screen.findByTestId(changeFileID)).resolves.toBeVisible();
	});

	it("should successfully import profile and return to home screen", async () => {
		const historyMock = vi.spyOn(history, "push").mockReturnValue();

		render(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(screen.getByTestId(changeFileID)).toBeInTheDocument();
		expect(screen.getByTestId("SelectFileStep__back")).toBeInTheDocument();

		fireEvent.drop(browseFiles(), {
			dataTransfer: {
				files: [createBlob(passwordProtectedWwe)],
			},
		});

		await waitFor(() => {
			expect(screen.getByTestId("ProcessingImport")).toBeVisible();
		});

		await expect(screen.findByTestId("Modal__inner")).resolves.toBeVisible();

		await userEvent.type(screen.getByTestId("PasswordModal__input"), validPassword);
		await waitFor(() => expect(screen.getByTestId("PasswordModal__input")).toHaveValue(validPassword));

		await userEvent.click(screen.getByTestId(submitID));

		await expect(screen.findByTestId("ProfileForm__form")).resolves.toBeVisible();

		expect(screen.queryByTestId("InputPassword")).not.toBeInTheDocument();

		await userEvent.click(screen.getByRole("checkbox"));

		await waitFor(() => {
			expect(screen.getByTestId("ProfileForm__submit-button")).toBeEnabled();
		});

		await userEvent.click(screen.getByTestId("ProfileForm__submit-button"));

		await waitFor(() => expect(historyMock).toHaveBeenCalledWith("/"));
	});

	it("should successfully import legacy profile and return to home screen", async () => {
		const historyMock = vi.spyOn(history, "push").mockReturnValue();

		render(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(screen.getByTestId(changeFileID)).toBeInTheDocument();
		expect(screen.getByTestId("SelectFileStep__back")).toBeInTheDocument();

		await userEvent.click(screen.getByTestId(changeFileID));

		fireEvent.drop(browseFiles(), {
			dataTransfer: {
				files: [createBlob(legacyJson, "legacy-profile.json")],
			},
		});

		await waitFor(() => {
			expect(screen.getByTestId("ProcessingImport")).toBeVisible();
		});

		await expect(screen.findByTestId("ProfileForm__form")).resolves.toBeVisible();

		await waitFor(() => expect(screen.getAllByTestId("Input")[0]).toHaveValue("legacy-profile"));

		await userEvent.click(screen.getByRole("checkbox"));

		await waitFor(() => {
			expect(screen.getByTestId("ProfileForm__submit-button")).toBeEnabled();
		});

		await userEvent.click(screen.getByTestId("ProfileForm__submit-button"));

		await waitFor(() => expect(historyMock).toHaveBeenCalledWith("/"));
	});

	it.each([
		["dark", darkThemeWwe],
		["light", lightThemeWwe],
	])("should apply theme setting of imported profile regardless of OS preferences", async (theme, wweFile) => {
		render(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(screen.getByTestId(changeFileID)).toBeInTheDocument();
		expect(screen.getByTestId("SelectFileStep__back")).toBeInTheDocument();

		fireEvent.drop(browseFiles(), {
			dataTransfer: {
				files: [createBlob(wweFile)],
			},
		});

		await waitFor(() => {
			expect(screen.getByTestId("ProcessingImport")).toBeVisible();
		});

		await expect(screen.findByTestId("ProfileForm__form")).resolves.toBeVisible();

		const lightButton = screen.getAllByRole("radio")[0];
		const darkButton = screen.getAllByRole("radio")[1];

		if (theme === "dark") {
			expect(lightButton).not.toBeChecked();
			expect(darkButton).toBeChecked();
		} else {
			expect(lightButton).toBeChecked();
			expect(darkButton).not.toBeChecked();
		}
	});

	it("should go to step 3 and back", async () => {
		render(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(screen.getByTestId(changeFileID)).toBeInTheDocument();
		expect(screen.getByTestId("SelectFileStep__back")).toBeInTheDocument();

		fireEvent.drop(browseFiles(), {
			dataTransfer: {
				files: [createBlob(passwordProtectedWwe)],
			},
		});

		await waitFor(() => {
			expect(screen.getByTestId("ProcessingImport")).toBeVisible();
		});

		await expect(screen.findByTestId("Modal__inner")).resolves.toBeVisible();

		await userEvent.type(screen.getByTestId("PasswordModal__input"), validPassword);

		await waitFor(() => {
			expect(screen.getByTestId("PasswordModal__input")).toHaveValue(validPassword);
		});

		await userEvent.click(screen.getByTestId(submitID));

		await expect(screen.findByTestId("ProfileForm__form")).resolves.toBeVisible();

		await userEvent.click(screen.getByTestId("ProfileForm__back-button"));

		await expect(screen.findByTestId(changeFileID)).resolves.toBeVisible();
	});

	it("should fail profile import and show error", async () => {
		render(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(screen.getByTestId(changeFileID)).toBeInTheDocument();
		expect(screen.getByTestId("SelectFileStep__back")).toBeInTheDocument();

		fireEvent.drop(browseFiles(), {
			dataTransfer: {
				files: [createBlob(corruptedWwe)],
			},
		});

		await expect(screen.findByTestId("ProcessingImport")).resolves.toBeVisible();
		await expect(screen.findByTestId("Modal__inner")).resolves.toBeVisible();

		await userEvent.type(screen.getByTestId("PasswordModal__input"), wrongPassword);

		await waitFor(() => {
			expect(screen.getByTestId("PasswordModal__input")).toHaveValue(wrongPassword);
		});

		await userEvent.click(screen.getByTestId(submitID));

		await expect(screen.findByTestId("ImportError")).resolves.toBeVisible();
	});

	it("should fail profile import and retry", async () => {
		render(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(screen.getByTestId(changeFileID)).toBeInTheDocument();
		expect(screen.getByTestId("SelectFileStep__back")).toBeInTheDocument();

		fireEvent.drop(browseFiles(), {
			dataTransfer: {
				files: [createBlob(passwordProtectedWwe)],
			},
		});

		await expect(screen.findByTestId("ProcessingImport")).resolves.toBeVisible();
		await expect(screen.findByTestId("Modal__inner")).resolves.toBeVisible();

		await userEvent.type(screen.getByTestId("PasswordModal__input"), wrongPassword);

		await waitFor(() => {
			expect(screen.getByTestId("PasswordModal__input")).toHaveValue(wrongPassword);
		});

		await userEvent.click(screen.getByTestId(submitID));

		await expect(screen.findByTestId("ImportError")).resolves.toBeVisible();

		await userEvent.click(screen.getByTestId("ImportError__retry"));

		await expect(screen.findByTestId("ProcessingImport")).resolves.toBeVisible();
	});

	it("should fail profile import and go back to home screen", async () => {
		const historyMock = vi.spyOn(history, "push").mockReturnValue();

		render(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(screen.getByTestId(changeFileID)).toBeInTheDocument();
		expect(screen.getByTestId("SelectFileStep__back")).toBeInTheDocument();

		fireEvent.drop(browseFiles(), {
			dataTransfer: {
				files: [createBlob(passwordProtectedWwe)],
			},
		});

		await expect(screen.findByTestId("ProcessingImport")).resolves.toBeVisible();
		await expect(screen.findByTestId("Modal__inner")).resolves.toBeVisible();

		await userEvent.type(screen.getByTestId("PasswordModal__input"), wrongPassword);

		await waitFor(() => {
			expect(screen.getByTestId("PasswordModal__input")).toHaveValue(wrongPassword);
		});

		await userEvent.click(screen.getByTestId(submitID));

		await expect(screen.findByTestId("ImportError")).resolves.toBeVisible();

		await userEvent.click(screen.getByTestId("ImportError__cancel"));

		await expect(screen.findByTestId("ImportError")).resolves.toBeVisible();

		await waitFor(() => expect(historyMock).toHaveBeenCalledWith("/"));

		historyMock.mockRestore();
	});
});
