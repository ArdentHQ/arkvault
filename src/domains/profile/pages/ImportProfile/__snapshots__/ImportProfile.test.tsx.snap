// Vitest Snapshot v1

exports[`ImportProfile > should render first step 1`] = `
<div>
  <div
    class="css-15789f"
  >
    <nav
      aria-labelledby="main menu"
      class="css-97i12m"
    >
      <div
        class="relative flex h-21"
      >
        <div
          class="flex flex-1 px-8 sm:ml-12"
        >
          <div
            class="my-auto flex items-center"
          >
            <button
              class="my-auto mr-4 flex h-11 w-11 cursor-pointer items-center justify-center rounded-xl bg-theme-primary-600 text-white outline-none focus:outline-none focus:ring-2 focus:ring-theme-primary-400"
              data-testid="NavigationBarLogo--button"
              type="button"
            >
              <span
                class="relative"
              >
                <svg
                  fill="none"
                  height="28"
                  viewBox="0 0 363 363"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g
                    clip-path="url(#clip0_434_4442)"
                  >
                    <path
                      d="M174.182 352.325a279.256 279.256 0 0 1-25.627-20.536c-8.518-7.525-17.006-15.66-25.229-24.178a470.15 470.15 0 0 1-29.234-33.315 449.417 449.417 0 0 1-57.68-93.623 420.933 420.933 0 0 1-22.044-60.102A416.702 416.702 0 0 1 1.126 51.749c-.512-4.803-.303-10.73 5.9-13.842a393.077 393.077 0 0 1 54.42-19.94 459.697 459.697 0 0 1 52.646-12.195A374.409 374.409 0 0 1 176.15.333h11.6a374.406 374.406 0 0 1 62.057 5.44 459.723 459.723 0 0 1 52.646 12.195 393.328 393.328 0 0 1 54.422 19.939c6.203 3.115 6.41 9.034 5.898 13.842a416.672 416.672 0 0 1-13.242 68.822 421.133 421.133 0 0 1-22.044 60.102 449.395 449.395 0 0 1-57.68 93.623 470 470 0 0 1-29.234 33.315c-8.223 8.518-16.711 16.653-25.229 24.178a279.73 279.73 0 0 1-25.627 20.536 17.811 17.811 0 0 1-15.528 0h-.007zm6.769-230.087c-.412.547-28.363 29.571-48.772 50.764-7.035 7.318-13.229 13.736-17.187 17.853a49.345 49.345 0 0 0-14.37 29.355 42.981 42.981 0 0 0 2.034 17.531 59.295 59.295 0 0 0 10.41 18.547 369.34 369.34 0 0 0 42.811 44.915 223.324 223.324 0 0 0 20.072 16.051 13.364 13.364 0 0 0 12.005 0 223.333 223.333 0 0 0 20.07-16.051 369.191 369.191 0 0 0 42.811-44.915 59.297 59.297 0 0 0 10.411-18.547 42.98 42.98 0 0 0 2.033-17.531 49.34 49.34 0 0 0-14.377-29.358l-14.811-15.385-1.779-1.848c-20.66-21.451-48.955-50.832-49.371-51.383a1.179 1.179 0 0 0-.995-.548 1.184 1.184 0 0 0-.995.548v.002zm1.408-69.483c.136.215.321.505.582.869.391.544 6.257 9.93 14.382 22.92 19.488 31.162 52.102 83.315 60.679 96.014 6.342 9.377 14.353 14.77 21.974 14.772 7.402 0 18.108-5.031 27.079-29.067a332.382 332.382 0 0 0 19.124-81.095c.409-3.84.243-8.568-4.713-11.056a314.32 314.32 0 0 0-43.468-15.926 367.237 367.237 0 0 0-42.052-9.743 299.13 299.13 0 0 0-49.568-4.345h-8.871a299.13 299.13 0 0 0-49.568 4.345 367.225 367.225 0 0 0-42.051 9.743 314.427 314.427 0 0 0-43.469 15.926c-4.951 2.486-5.12 7.213-4.712 11.056a332.313 332.313 0 0 0 19.124 81.095c4.952 13.265 13.67 29.067 27.078 29.067 7.624 0 15.63-5.384 21.974-14.772 8.57-12.687 41.169-64.811 60.643-95.955l.032-.053c8.124-12.993 13.995-22.38 14.384-22.921.259-.364.455-.657.58-.87.209-.327.311-.491.416-.491.105 0 .214.16.421.487zm-55.455 149.533l19.652-20.614h69.159l19.649 20.614h-108.46zm34.197-35.377l20.162-20.871 20.159 20.871h-40.321z"
                      fill="currentColor"
                    />
                  </g>
                </svg>
              </span>
            </button>
            <span
              class="text-lg uppercase"
            >
              <strong>
                ARK
              </strong>
              Vault
            </span>
          </div>
        </div>
      </div>
    </nav>
    <div
      class="flex flex-1 flex-col"
    >
      <div
        class="-mt-8 css-ycib48"
      >
        <div
          class="mx-auto px-8 md:px-10 lg:container"
        >
          <div
            class="mt-6 sm:mt-8"
          >
            <div
              data-testid="tab-pabel__active-panel"
            >
              <div
                class="mx-auto max-w-xl"
              >
                <div
                  class="sm:mt-8"
                >
                  <div
                    class="flex flex-col"
                  >
                    <span
                      class="css-3qwsx5"
                    >
                      Import Profile
                    </span>
                    <ul
                      class="css-xa5o6q"
                    >
                      <li
                        class="css-47028"
                      />
                      <li
                        class="css-3ldhk3"
                      />
                      <li
                        class="css-3ldhk3"
                      />
                    </ul>
                  </div>
                  <div
                    class="flex flex-col justify-between sm:flex-row"
                  >
                    <div
                      class="flex items-end justify-between bg-theme-background mt-6 hidden sm:block"
                    >
                      <div
                        class="space-y-4"
                      >
                        <h1
                          class="mb-0 text-2xl"
                          data-testid="header__title"
                        >
                          Import Profile
                        </h1>
                        <div
                          class="flex items-center text-theme-secondary-text"
                          data-testid="header__subtitle"
                        >
                          Select a .wwe file with your Profile and related properties to start the import process.
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="relative mt-6 h-52 rounded-lg border-2 border-dashed border-theme-secondary-300 p-2 dark:border-theme-secondary-800 sm:mt-8"
                  data-testid="SelectFile"
                >
                  <div
                    class="flex h-full flex-col items-center justify-center rounded-lg transition-colors duration-200 bg-theme-primary-50 dark:bg-theme-secondary-800"
                    data-testid="SelectFile__drop-zone"
                  >
                    <div
                      class="css-1atxdzc"
                      height="40"
                      width="40"
                    >
                      <svg
                        id="extension-wwe"
                        viewBox="0 0 40 40"
                        x="0"
                        xml:space="preserve"
                        xmlns="http://www.w3.org/2000/svg"
                        y="0"
                      >
                        <path
                          d="M7.757 37H3.598C2.17 37 1 35.898 1 34.55V3.45C1 2.101 2.17 1 3.469 1h17.929c.78 0 1.298.245 1.818.735l10.004 9.428c.52.49.78 1.102.78 1.715v7.347"
                          fill="none"
                          stroke="#289548"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="2"
                        />
                        <g
                          fill="none"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="2"
                        >
                          <path
                            d="M22 29v3.396c0 2.35.696 4.648 2 6.604 1.31-1.978 2.004-3.664 2-6"
                          />
                          <path
                            d="M26 33c0 2.35.697 4.044 2 5.999 1.304-1.956 2-4.253 2-6.604V29m7 10h-2a2.002 2.002 0 0 1-2-2.001v-6A2 2 0 0 1 35 29h2m-4 6h4m-26-6v3.396c0 2.35.696 4.648 2 6.604 1.31-1.978 2.004-3.664 2-6"
                          />
                          <path
                            d="M15 33c0 2.35.697 4.044 2 5.999 1.304-1.956 2-4.254 2-6.604V29"
                          />
                        </g>
                      </svg>
                    </div>
                    <div
                      class="mt-8"
                    >
                      <span
                        class="mr-px hidden font-semibold sm:inline"
                      >
                        Drag & Drop or
                         
                      </span>
                      <button
                        class="link ring-focus relative cursor-pointer font-semibold focus:outline-none"
                        data-ring-focus-margin="-m-1"
                        data-testid="SelectFile__browse-files"
                        title="Click here to upload"
                        type="button"
                      >
                        Browse Files
                      </button>
                    </div>
                    <div
                      class="mt-2 text-sm font-semibold text-theme-secondary-500"
                    >
                      Supported format is .wwe
                    </div>
                  </div>
                </div>
                <p
                  class="mt-6 text-center text-base text-theme-secondary-text"
                >
                  <span>
                    Importing from Desktop Wallet v2?
                     
                  </span>
                  <button
                    class="link ring-focus relative cursor-pointer font-semibold focus:outline-none"
                    data-ring-focus-margin="-m-1"
                    data-testid="SelectFileStep__change-file"
                    title="Click here."
                    type="button"
                  >
                    Click here.
                  </button>
                </p>
                <div
                  class="css-1tkbsa1"
                >
                  <button
                    class="css-1jhy6v9"
                    data-testid="SelectFileStep__back"
                    type="button"
                  >
                    <div
                      class="flex items-center space-x-2"
                    >
                      Back
                    </div>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
